nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-kubernetes-bundle/nifi-framework-kubernetes-state-provider/src/main/java/org/apache/nifi/kubernetes/state/provider/KubernetesConfigMapStateProvider.java

```diff
@@ -0,0 +1,317 @@
+        } catch (final RuntimeException e) {
+            throw new IOException(String.format("Clear failed for Component ID [%s]", componentId), e);
+        }
+    }
+
+    /**
+     * Remove state information for specified Component Identifier
+     *
+     * @param componentId Identifier of component removed from the configuration
+     * @throws IOException Thrown on failure to remove state for Component Identifier
+     */
+    @Override
+    public void onComponentRemoved(final String componentId) throws IOException {
+        try {
+            final List<StatusDetails> deleteStatus = configMapResource(componentId).delete();
+            logger.debug("Config Map [{}] deleted {}", componentId, deleteStatus);
+        } catch (final RuntimeException e) {
+            throw new IOException(String.format("Remove failed for Component ID [%s]", componentId), e);
+        }
+    }
+
+    /**
+     * Enable Provider
+     */
+    @Override
+    public void enable() {
+        enabled.getAndSet(true);
+    }
+
+    /**
+     * Disable Provider
+     */
+    @Override
+    public void disable() {
+        enabled.getAndSet(false);
+    }
+
+    /**
+     * Get Enabled status
+     *
+     * @return Enabled status
+     */
+    @Override
+    public boolean isEnabled() {
+        return enabled.get();
+    }
+
+    /**
+     * Get Supported Scopes returns CLUSTER
+     *
+     * @return Supported Scopes including CLUSTER
+     */
+    @Override
+    public Scope[] getSupportedScopes() {
+        return SUPPORTED_SCOPES;
+    }
+
+    /**
+     * Get Kubernetes Client using standard configuration
+     *
+     * @return Kubernetes Client
+     */
+    protected KubernetesClient getKubernetesClient() {
+        return new StandardKubernetesClientProvider().getKubernetesClient();
+    }
+
+    private Resource<ConfigMap> configMapResource(final String componentId) {
+        final String name = getConfigMapName(componentId);
+        return kubernetesClient.configMaps().inNamespace(namespace).withName(name);
+    }
+
+    private ConfigMapBuilder createConfigMapBuilder(final Map<String, String> state, final String componentId) {
+        final Map<String, String> encodedData = getEncodedMap(state);
+        final String name = getConfigMapName(componentId);
+        return new ConfigMapBuilder()
+                .withNewMetadata()
+                .withNamespace(namespace)
+                .withName(name)
+                .endMetadata()
+                .withData(encodedData);
+    }
+
+    private String getConfigMapName(final String componentId) {
+        return String.format(CONFIG_MAP_NAME_FORMAT, componentId);
+    }
+
+    private long getVersion(final ConfigMap configMap) {
+        final ObjectMeta metadata = configMap.getMetadata();
+        final String resourceVersion = metadata.getResourceVersion();
+        try {
+            return resourceVersion == null ? UNKNOWN_VERSION : Long.parseLong(resourceVersion);
```

Comment:
I don't believe this is valid. According to https://kubernetes.io/docs/reference/using-api/api-concepts/
```
You must not assume resource versions are numeric or collatable. API clients may only compare two resource versions for equality (this means that you must not compare resource versions for greater-than or less-than relationships).
```

Reply:
I think I would propose that we update our StateMap as follows:
Make `long getVersion()` deprecated in favor of a new `String getStateVersion()`
Remove `long getVersion()` in version 2.0

Reply:
I was wondering about this approach, it probably depends on the Kubernetes implementation. One option that comes to mind is creating using a hash of the Resource Version. Another option could be using a custom metadata field.

Reply:
Everywhere in the code base that I see `long getVersion()` being used, actually, is to check if the value is `-1` (indicating that the state hasn't actually been stored anywhere). So it probably actually makes sense to use an `Optional<String>` as the return type.

Reply:
Thanks for the suggestion, adding a new `Optional<String> getStateVersion()` method to `StateMap` sounds like a good approach.